site/content/contribute/server/style-guide.md

```diff
@@ -0,0 +1,98 @@
+---
+title: "Style guide"
+date: 2021-01-12T16:00:00+0530
+weight: 3
+subsection: Server
+---
+
+The overarching guideline is to just follow [Effective Go](https://golang.org/doc/effective_go.html) along with [CodeReviewComments](https://github.com/golang/go/wiki/CodeReviewComments). But currently, implementing those rules can lead to a clash between maintaining consistency and enforcing correctness. This document aims to codify some of the rules in the hope that any confusion is alleviated.
+
+### Model package
+
+The `model` package is our public API which is consumed by multiple plugins. Having a non-idiomatic public API should not prevent us from improving the rest of the codebase. More code keeps getting added, and it becomes practically impossible to fix all of the issues at once. Instead, we choose to have consistency in that package only if changing it makes it inconsistent from an API perspective. This includes both breaking changes and any inconsistencies that might appear in godocs.
+
+There are some rules which don’t need an API change and can be used everywhere. In all other packages, we should enforce correctness. Once we move to v6 of the `mattermost-server`, we can change the remaining places to match with the rest of the codebase.
+
+This does not, however, mean that a developer has to fix any surrounding code to fix the style issues too. It’s encouraged to keep fixing things as we go, but it's not compulsory to do so. There's also no need for a reviewer to ask for that change if it's not something that's added/edited in that PR.
+
+The following goes into some of the specific anti-patterns that have crept in our codebase and we should aim to correct. The rules which are exempt from being applied in the model package are specifically indicated.
+
+## Guidelines
+
+### Default to sync instead of async
+
+Always prefer synchronous functions by default. Async calls are hard to get right. They have no control over goroutine lifetimes and introduce data races. If you think something needs to be asynchronous, measure it and prove it. Ask these questions:
+- Does it improve performance? If so, by how much?
+- What’s the tradeoff of the happy path vs. slow path?
+- How do I propagate errors?
+- What about back-pressure?
+- What should be my concurrency model?
+
+Avoid creating one-off goroutines without knowing when/how they exit. They cause problems that are hard to debug, and can often cause performance degradation rather than an improvement. Have a look at:
+- https://github.com/golang/go/wiki/CodeReviewComments#goroutine-lifetimes
+- https://github.com/golang/go/wiki/CodeReviewComments#synchronous-functions
+
+### Pointers to slices (exempt from model)
+
+Avoid using pointers to slices. Slices are already reference types which point to an underlying array. If you want a function to modify a slice, then return that slice from the function, rather than passing a pointer.
```

Comment:
```suggestion
Do not use pointers to slices. Slices are already reference types which point to an underlying array. If you want a function to modify a slice, then return that slice from the function, rather than passing a pointer.
```